lightning/src/util/persist.rs

```diff
@@ -64,41 +114,821 @@ impl<'a, A: KVStorePersister, M: Deref, T: Deref, ES: Deref, NS: Deref, SP: Dere
+		check_persisted_data!(5);
+		send_payment(&nodes[1], &vec![&nodes[0]][..], 4000000);
+		check_persisted_data!(10);
+
+		// Force close because cooperative close doesn't result in any persisted
+		// updates.
+		nodes[0]
+			.node
+			.force_close_broadcasting_latest_txn(
+				&nodes[0].node.list_channels()[0].channel_id,
+				&nodes[1].node.get_our_node_id(),
+			)
+			.unwrap();
+
+		check_closed_event(&nodes[0], 1, ClosureReason::HolderForceClosed, false, &[nodes[1].node.get_our_node_id()], 100000);
+		check_closed_broadcast!(nodes[0], true);
+		check_added_monitors!(nodes[0], 1);
+
+		let node_txn = nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap();
+		assert_eq!(node_txn.len(), 1);
+
+		connect_block(
+			&nodes[1],
+			&create_dummy_block(
+				nodes[0].best_block_hash(),
+				42,
+				vec![node_txn[0].clone(), node_txn[0].clone()],
+			),
+		);
+		check_closed_broadcast!(nodes[1], true);
+		check_closed_event(&nodes[1], 1, ClosureReason::CommitmentTxConfirmed, false, &[nodes[0].node.get_our_node_id()], 100000);
+		check_added_monitors!(nodes[1], 1);
+
+		// Make sure everything is persisted as expected after close.
+		check_persisted_data!(CLOSED_CHANNEL_UPDATE_ID);
+	}
+
+	// Test that if the persister's path to channel data is read-only, writing a
+	// monitor to it results in the persister returning a PermanentFailure.
+	#[test]
+	fn test_ro_device_perm_failure() {
+		// Set up a dummy channel and force close. This will produce a monitor
+		// that we can then use to test persistence.
+		let chanmon_cfgs = create_chanmon_cfgs(2);
+		let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+		let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+		let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+		let chan = create_announced_chan_between_nodes(&nodes, 0, 1);
+		nodes[1].node.force_close_broadcasting_latest_txn(&chan.2, &nodes[0].node.get_our_node_id()).unwrap();
+		check_closed_event(&nodes[1], 1, ClosureReason::HolderForceClosed, false, &[nodes[0].node.get_our_node_id()], 100000);
+		let mut added_monitors = nodes[1].chain_monitor.added_monitors.lock().unwrap();
+		let update_map = nodes[1].chain_monitor.latest_monitor_update_id.lock().unwrap();
+		let update_id = update_map.get(&added_monitors[0].0.to_channel_id()).unwrap();
+		let test_txo = OutPoint {
+			txid: Txid::from_hex(
+				"8984484a580b825b9972d7adb15050b3ab624ccd731946b3eeddb92f4e7ef6be",
+			)
+			.unwrap(),
+			index: 0,
+		};
+
+		let persister = read_only_test_persister();
+		match persister.persist_new_channel(test_txo, &added_monitors[0].1, update_id.2) {
+			ChannelMonitorUpdateStatus::PermanentFailure => {
+				// Correct result
+			}
+			ChannelMonitorUpdateStatus::Completed => {
+				panic!("Completed persisting new channel when shouldn't have")
+			}
+			ChannelMonitorUpdateStatus::InProgress => {
+				panic!("Started persisting new channel when shouldn't have")
+			}
+		}
+		nodes[1].node.get_and_clear_pending_msg_events();
```

Comment:
I think I could use some help with this test. I'm fairly sure the actual subject of the test works -- the `match` statement above completes on the correct branch. But there's a panic that I think comes from this line.

```thread 'util::persist::tests::test_ro_device_perm_failure' panicked at 'Tried to violate existing lockorder. Build with the backtrace feature for more info.', lightning/src/sync/debug_sync.rs:179:29```

Nothing obvious in the backtrace I have.

Reply:
nm, figured out with some help that the lock on added monitors had to be dropped